### ISSUE TITLE
Deprecate twisted.words.protocols.msn

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6395]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#6395|
|Type|task|
|Created|2013-03-28 13:24:35Z|
|Branch|https://github.com/twisted/twisted/tree/deprecate-msn-6395-2|

MSN is shutting down the service, I don't think the client code has any usefulness by itself. We should remove the associated example too.

Attachments:

* [deleted_msn.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cc9/cc95db3089ae20e1e1d0185f968d06ab8dd5e906/c77640eb581f84b80d4148c64442da55f7d8eed3.patch) (27 bytes) - added by dhanush on 2013-03-30 20:33:22Z - Deleted twisted.words.protocols.msn
* [deprecated_msn.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cc9/cc95db3089ae20e1e1d0185f968d06ab8dd5e906/084bedb3e91df537aa4e1c4b25a8a31148ece645.patch) (430 bytes) - added by dhanush on 2013-04-01 18:55:35Z - Deprecated twisted.words.protocols.msn
* [deprecated_msn_added_corrections.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cc9/cc95db3089ae20e1e1d0185f968d06ab8dd5e906/4d2f59119f249caac603b0e9e9f2ed4d7acad03b.patch) (2306 bytes) - added by dhanush on 2013-04-07 15:49:13Z - Added NEWS File and tests to deprecate twisted.words.protocols.msn
* [deprecated_msn_added_corrections_2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cc9/cc95db3089ae20e1e1d0185f968d06ab8dd5e906/eb5fa6092294f22545c5c95783a012674d4215c2.patch) (2633 bytes) - added by dhanush on 2013-04-17 10:39:21Z - Made changes according to review points. But trial twisted generates 2 errors.
* [deprecated_msn_added_corrections_2.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cc9/cc95db3089ae20e1e1d0185f968d06ab8dd5e906/6094e22efb2a5688234990508785662ed6d7a118.patch) (2633 bytes) - added by dhanush on 2013-04-17 10:39:48Z - Made changes according to review points. But trial twisted generates 2 errors.
* [6395.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cc9/cc95db3089ae20e1e1d0185f968d06ab8dd5e906/51b67cc8a823373d674e10d51f955b586eccda60.patch) (2140 bytes) - added by moijes12 on 2014-04-22 17:46:13Z - Completing dhanush's work. Tests have been run against the patch and they pass. Also, the warning is now raised.
* [6395-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cc9/cc95db3089ae20e1e1d0185f968d06ab8dd5e906/67dcf4a5bfb41a450446ff24fd31e3c362e2065a.patch) (4466 bytes) - added by moijes12 on 2014-04-22 19:33:32Z - Building on thijs' comments. Attaching a tested patch. 

<details><summary>Searchable metadata</summary>

```
trac-id__6395 6395
type__task task
reporter__therve therve
priority__normal normal
milestone__ 
branch__branches_deprecate_msn_6395_2 branches/deprecate-msn-6395-2
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__words words
keywords__easy easy
time__1364477075000000 1364477075000000
changetime__1422200800173812 1422200800173812
version__None None
owner__hawkowl hawkowl
cc__ralphm cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

Hi your patch seems to be incorrect. Why is this breaking the compatibility policy by not deprecating it first?

**Comment by twisted-trac** [comment]
|<img alt="dhanush's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|dhanush commented|
|-|-|

Yes you are right. Sorry I was unaware about this. Can you tell me how to deprecate a module? I've looked up in http://twistedmatrix.com/trac/browser/tags/releases/twisted-12.3.0/twisted/python/deprecate.py
But didn't make out much out  of it.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst set owner to dhanush|
|-|-|

Thanks for the patch!

  1. The comment is unnecessary, it's pretty clear what's going on.
  2. Please run diff at the root of the Twisted checkout, so it's clear which file the patch applies to.
  3. All code changes to Twisted need tests. `twisted/conch/test/test_insults.py` should provide an example of how to test deprecation of modules.
  4. The patch should include a news file (https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles).

Please address these points and resubmit your patch.

**Comment by twisted-trac** [comment]
|<img alt="dhanush's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|dhanush commented|
|-|-|

Applied the corrections.

**Comment by twisted-trac** [comment]
|<img alt="dhanush's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|dhanush removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_11"></a>|@tomprince set owner to dhanush|
|-|-|

1.Did you try running the tests? The code fails to run for me. (But see below)
 2. There isn't a warning emitted 
```
#!python
from twisted.words.protocols.msn import NotificationClient
```
    but there probably should be. `twisted/protocols/telnet.py` has an example of how to emit a deprecation warning that handles both that and
```
#!python
from twisted.words.protocols import msn
```
    You should have a test for both cases.
 3. Calls to `flushWarnings` should pass the function that is to be blamed by the warning (in this case `test_msn`.
    - Incidentally, that test should probably be something like `test_msnDeprecated`, and the docstring could be more specific: `When importing .... a L{DeprecationWarning} is reported`. ([#6301](https://github.com/twisted/twisted/issues/6301) has some comments on how to write test docstrings)

**Comment by twisted-trac** [comment]
|<img alt="dhanush's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|dhanush commented|
|-|-|

Made some changes according to suggestions. But trial twisted generates the following errors:


```
[ERROR]
Traceback (most recent call last):
  File "/home/dhanush/Twisted 12.3.0-original/twisted/mail/test/test_mail.py", line 1937, in _cbProcessAlias
    lines = file('process.alias.out').readlines()
exceptions.IOError: [Errno 2] No such file or directory: 'process.alias.out'

twisted.mail.test.test_mail.ProcessAliasTestCase.test_processAlias
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/dhanush/Twisted 12.3.0-original/twisted/words/test/test_msn.py", line 545, in test_msnDeprecated
    self.ensureDeprecated("twisted.words.protocols.msn is deprecated "
  File "/home/dhanush/Twisted 12.3.0-original/twisted/words/test/test_msn.py", line 535, in ensureDeprecated
    self.assertIdentical(warnings[0]['category'], DeprecationWarning)
exceptions.IndexError: list index out of range

twisted.words.test.test_msn.Deprecations.test_msnDeprecated

```

Error 1 looks to be unrelated to this patch. Couldn't resolve Error 2. Awaiting some advice.

**Comment by twisted-trac** [comment]
|<img alt="dhanush's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|dhanush removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="dhanush's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|dhanush set owner to @tomprince|
|-|-|

I am interested in participating in GSoC 13 under Twisted. itamar suggested to assign it to you.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_15"></a>|@tomprince set owner to dhanush|
|-|-|

1.The first error is unrelated to this ticket, and is due to the space in your path. (There is already a ticket about addressing this)
   - Based on the path I see there, it also appears that you are working against a tarball from 12.3. If you are planning on contributing to twisted, you should really be working from a source checkout see TwistedDevelopment).
 2. The error you are getting is because there wasn't a warning emitted *during* the test. This is because the module has already been imported, and so the the import statment isn't actually running the code that triggers the deprecation warning.
   - As therve suggested on irc, have a look at `twisted/internet/test/test_tls.py` for an example of how to ensure that the code triggering the warning gets run.
   - Putting the `assertEqual(len(warnings), 1)` at the begining, rather than the end, might have made the error easier to understand: it would have said that the `len(warnings)` is 0 not 1, rather then having to realise that `IndexError` means that `warnings` isnt long enough.

**Comment by twisted-trac** [comment]
|[<img alt="moijes12's avatar" src="https://avatars.githubusercontent.com/u/4566851?s=50" width="50" height="50">](https://github.com/moijes12)|@moijes12 removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_17"></a>|@thijstriemstra set owner to @moijes12|
|-|-|

Thanks for your patch.

 1. the version nr should be bumped to 14.1
 1. there's still an example in [docs/words/examples/msn_example.py](docs/words/examples/msn_example.py) that needs to be removed

**Comment by twisted-trac** [comment]
|[<img alt="moijes12's avatar" src="https://avatars.githubusercontent.com/u/4566851?s=50" width="50" height="50">](https://github.com/moijes12)|@moijes12 removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl set owner to @hawkowl|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_20"></a>|@hawkowl commented|
|-|-|

(In [43363]) Branching to deprecate-msn-6395.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_21"></a>|@hawkowl commented|
|-|-|

(In [43364]) applying patch from moijes12, refs [#6395](https://github.com/twisted/twisted/issues/6395)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_22"></a>|@hawkowl commented|
|-|-|

LGTM. I'll merge it tomorrow.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

> there's still an example in trunk/docs/words/examples/msn_example.py​ that needs to be removed

This is incorrect.  The standing policy is to remove the documentation when the feature is removed.  Not to remove the documentation when the feature is deprecated.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_24"></a>|@hawkowl commented|
|-|-|

Replying to [exarkun](#note_23):
> 
> This is incorrect.  The standing policy is to remove the documentation when the feature is removed.  Not to remove the documentation when the feature is deprecated.

I didn't know this was part of policy, and I can't find it in the Documentation standard or Compatibility Policy -- should we add it, or is it somewhere else?

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_25"></a>|@tomprince commented|
|-|-|

Replying to [hawkowl](#note_24):
> Replying to [exarkun](#note_23):
> > 
> > This is incorrect.  The standing policy is to remove the documentation when the feature is removed.  Not to remove the documentation when the feature is deprecated.
> 
> I didn't know this was part of policy, and I can't find it in the Documentation standard or Compatibility Policy -- should we add it, or is it somewhere else?

I'm not sure if it is written down anywhere (if it isn't it should be). The reasoning is, that until the code is gone, people may be using it and so we should keep existing documentation about it.

 - Perhaps they inherited a codebase that is already using it, and they need to figure it out before removing the usage
 - In a case like this, where we are removing functionality, they may have use of it, even if we aren't going to continue supporting it.

In general, the twisted policy is to keep code around as long as it isn't a maintenance burden. Applying this to documentation, there is no reason to remove documentation until the associated code is removed. Once the code is gone, it isn't useful, but it may potentially be, up to that point.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_26"></a>|@hawkowl commented|
|-|-|

Replying to [tom.prince](#note_25):
> Replying to [hawkowl](#note_24):
> > Replying to [exarkun](#note_23):
> > > 
> > > This is incorrect.  The standing policy is to remove the documentation when the feature is removed.  Not to remove the documentation when the feature is deprecated.
> > 
> > I didn't know this was part of policy, and I can't find it in the Documentation standard or Compatibility Policy -- should we add it, or is it somewhere else?
> 
> I'm not sure if it is written down anywhere (if it isn't it should be). The reasoning is, that until the code is gone, people may be using it and so we should keep existing documentation about it.
> 
>  - Perhaps they inherited a codebase that is already using it, and they need to figure it out before removing the usage
>  - In a case like this, where we are removing functionality, they may have use of it, even if we aren't going to continue supporting it.
> 
> In general, the twisted policy is to keep code around as long as it isn't a maintenance burden. Applying this to documentation, there is no reason to remove documentation until the associated code is removed. Once the code is gone, it isn't useful, but it may potentially be, up to that point.

Right. This is an interesting case, because that example won't work (the endpoint is gone), and I'm not sure it could even work (does a 3rd party MSN server exist?). So, I'm not sure if it's really worth keeping, because it is essentially unhelpful.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_27"></a>|@hawkowl commented|
|-|-|

It turns out MSN *isn't dead*. http://ismsndeadyet.com/ With the correct IP, the example can still work. So I'll do a branch that is pre-that.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun commented|
|-|-|

> Right. This is an interesting case, because that example won't work (the endpoint is gone), and I'm not sure it could even work (does a 3rd party MSN server exist?). So, I'm not sure if it's really worth keeping, because it is essentially unhelpful.

Or you could just do what I asked instead of trying to invent special cases and move on from this fairly irrelevant piece of work to more useful things.

Heck, if that had happened 22 months ago we could be *removing* the MSN implementation instead of still discussing how to deprecate it.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_29"></a>|@hawkowl commented|
|-|-|

(In [43684]) Branching to deprecate-msn-6395-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_30"></a>|@hawkowl commented|
|-|-|

(In [43685]) applying patch from moijes12, refs [#6395](https://github.com/twisted/twisted/issues/6395) + moving it to 15.1

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_31"></a>|@hawkowl removed owner|
|-|-|

Moved the topfile (didn't notice that last time), shifted the version numbers to 15.1, and forced a build on the buildbots. Since I modified it, I'll put it up for review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_32"></a>|@adiroiban commented|
|-|-|

I have fixed the pyflakes test and triggers a new round of tests.

Hope all is ok.

Otherwise the changes looks good.

Not sure if I have the right to merge this as I have also made a change... but the change is minor.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_33"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Thanks for the work on this.

All look good!

I have fixed the pyflakes error, but on the last buildbot run one test is failing on win xp... but I think that this is not related.

```
twisted.words.test.test_xpath.XPathTests.test_position
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "C:\buildslave\twisted\winxp32-py2.7\Twisted\twisted\trial\_dist\test\test_disttrial.py", line 371, in test_runUntilFailure
    self.assertEqual(5, len(called))
  File "C:\buildslave\twisted\winxp32-py2.7\Twisted\twisted\trial\_synctest.py", line 447, in assertEqual
    % (msg, pformat(first), pformat(second)))
twisted.trial.unittest.FailTest: not equal:
a = 5
b = 2
```

Please take a look and see if this is spurious or no... if is spurious maybe we should create a ticket and fix it. I am happy to work on that as I dislike having spurious tests.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_34"></a>|@hawkowl set status to `closed`|
|-|-|

(In [43763]) Merge deprecate-msn-6395-2: Deprecate MSN support

Authors: dhanush, moijes12, adiroiban
Reviewers: itamar, tom.prince, thijs, hawkowl, adiroiban
Fixes: [#6395](https://github.com/twisted/twisted/issues/6395)
